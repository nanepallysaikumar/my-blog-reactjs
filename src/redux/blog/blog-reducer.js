import {
    BLOG_UPDATE_LIST,
    BLOG_ADD_ITEM,
    BLOG_REMOVE_ITEM,
    BLOG_EDIT_ITEM
} from './blog-constants';

const initialState = {
    list: []
};

function editBlogItem(state, item) {
    const index = state.list.findIndex(blog => blog.id === item.id);
    if (index === -1) {
        return state;
    }
    return {
        list: [
            ...state.list.slice(0, index),
            { ...item },
            ...state.list.slice(index + 1)
        ]
    };
}

function removeBlogItem(state, id) {
    const index = state.list.findIndex(blog => blog.id === id);
    if (index === -1) {
        return state;
    }
    return {
        list: [...state.list.slice(0, index), ...state.list.slice(index + 1)]
    };
}

export function blog(state = initialState, action) {
    switch (action.type) {
        case BLOG_UPDATE_LIST:
            return {
                list: [...action.payload]
            };

        case BLOG_ADD_ITEM:
            return {
                list: [...state.list, action.payload]
            };

        case BLOG_EDIT_ITEM:
            return editBlogItem(state, action.payload);

        case BLOG_REMOVE_ITEM:
            return removeBlogItem(state, action.payload.id);

        default:
            return state;
    }
}
