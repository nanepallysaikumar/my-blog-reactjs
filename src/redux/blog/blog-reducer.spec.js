import { blog } from './blog-reducer';
import {
    BLOG_UPDATE_LIST,
    BLOG_ADD_ITEM,
    BLOG_EDIT_ITEM,
    BLOG_REMOVE_ITEM
} from './blog-constants';

describe('blog-reducer', () => {
    const state = {
        list: []
    };

    const data = {
        id: '12122',
        header: 'test header',
        body: 'test body'
    };

    it('should return current state on unknown action', () => {
        const nextState = blog(state, {});
        expect(nextState).toBe(state);
    });

    it('should update the list on BLOG_UPDATE_LIST action', () => {
        const nextState = blog(state, {
            type: BLOG_UPDATE_LIST,
            payload: [data]
        });
        expect(nextState).toMatchSnapshot('updateListSnapshot');
    });

    it('should add item on  BLOG_ADD_ITEM action', () => {
        const nextState = blog(state, {
            type: BLOG_ADD_ITEM,
            payload: data
        });
        expect(nextState).toMatchSnapshot('addItemSnapshot');
    });

    it('should update item on  BLOG_ADD_ITEM action', () => {
        const state = {
            list: [data]
        };
        const nextState = blog(state, {
            type: BLOG_EDIT_ITEM,
            payload: {
                id: data.id,
                header: 'new heading',
                body: 'new body'
            }
        });
        expect(nextState).toMatchSnapshot('updateItemSnapshot');
    });

    it('should remove item on BLOG_ADD_ITEM action', () => {
        const state = {
            list: [data]
        };
        const nextState = blog(state, {
            type: BLOG_REMOVE_ITEM,
            payload: data
        });
        expect(nextState.list.length).toBe(0);
    });
});
