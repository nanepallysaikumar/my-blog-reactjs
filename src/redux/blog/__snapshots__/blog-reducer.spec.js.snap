// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blog-reducer should add item on  BLOG_ADD_ITEM action: addItemSnapshot 1`] = `
Object {
  "list": Array [
    Object {
      "body": "test body",
      "header": "test header",
      "id": "12122",
    },
  ],
}
`;

exports[`blog-reducer should update item on  BLOG_ADD_ITEM action: updateItemSnapshot 1`] = `
Object {
  "list": Array [
    Object {
      "body": "new body",
      "header": "new heading",
      "id": "12122",
    },
  ],
}
`;

exports[`blog-reducer should update the list on BLOG_UPDATE_LIST action: updateListSnapshot 1`] = `
Object {
  "list": Array [
    Object {
      "body": "test body",
      "header": "test header",
      "id": "12122",
    },
  ],
}
`;
